db/db_impl/db_impl.cc

```diff
@@ -626,6 +626,19 @@ Status DBImpl::CloseHelper() {
     job_context.Clean();
     mutex_.Lock();
   }
+  if (!mutable_db_options_.avoid_flush_during_shutdown && !logs_.empty()) {
+    mutex_.Unlock();
+    Status s = SyncWAL();
```

Comment:
Looking for some ideas to benchmark @ajkr @ltamasi: since now DB Close() with WAL sync will be slower, how should we answer the question of "how much slower?".

I'm thinking of 
(1) Give a rough estimate of per sync time in OS ... probably googling for this number? And tell people to time this number by your #WAL files.

(2) Similar to (1) but microbench SyncWAL() on 1 WAL for estimating the per sync time

(3) ...?

Reply:
> how should we answer the question of "how much slower?"

IMO, "a lot slower" plus a reminder that the old behavior to not guarantee durability is available via `avoid_flush_during_shutdown == true`.

On the topic of durability violations, now `avoid_flush_during_shutdown == true` both makes WAL-enabled writes unsynced and drops WAL-disabled writes. I thought about the  possible use cases (especially ones that mix WAL-enabled/disabled writes) and think the risk of bundling these violations into one option is low, but I could be wrong.

Reply:
If you wanted the risk to be reduced you could introduce `avoid_sync_during_shutdown`. It would allow the following (hypothetical, AFAIK) use case to get the old behavior: (1) sometimes writes with WAL disabled, (2) must not lose WAL-disabled writes on clean shutdown, and (3) does not want to pay WAL sync overhead. It's kind of a confusing use case because (2) implies they are willing to pay SST sync overhead thus making (3) sound contradictory. But that is the case I thought of and the struggle to make sense out of it is why I thought the risk is low.

Reply:
> I thought the risk is low.

... With respect to the bundling durability violations into one option. Regarding the whole thing about changing the default behavior, that feels more like medium.

Reply:
Okay - now i'm more toward introducing avoid_sync_during_shutdown

Reply:
Ok but we will need to reserve time to actually follow through with users to get them to migrate / eventually change the default. Otherwise it'll be confusing for years to come regarding why we built a database without durability by default.

Reply:
Sounds good - will do so particularly with avoid_flush_during_recovery=false users

Reply:
It seems we can create a helper for SyncWAL() where the boolean parameter indicates whether lock on mutex_ has been taken.

This way we don't need to release / reacquire the lock.